Realm/RLMSyncUser.mm

```diff
@@ -391,130 +385,197 @@ + (void)confirmEmailForAuthServer:(NSURL *)serverURL
+     JSON:@{kRLMSyncTokenKey: self.refreshToken}
+     timeout:60.0
+     completion:^(NSError *error, NSDictionary *json) {
+        if (error) {
+            return callback(nil, error);
+        }
+        // FIXME: ROS currently gives duplicated results for 'all' due to an incorrect query
+        NSMutableSet *permissions = [NSMutableSet new];
+        for (NSDictionary *permission in json[@"permissions"]) {
+            // ROS reports the permission for __wildcardpermissions, which we
+            // don't want to include
+            if ([permission[@"path"] hasPrefix:@"/__"]) {
+                continue;
+            }
+
+            // Wildcard permissions are reported as a null userId
+            id userId = permission[@"userId"];
+            if (userId == NSNull.null) {
+                userId = @"*";
+            }
+
+            [permissions addObject:[[RLMSyncPermission alloc]
+                                    initWithRealmPath:permission[@"path"]
+                                    identity:userId
+                                    accessLevel:RLMSyncAccessLevelFromString(permission[@"accessLevel"])]];
+        }
+        callback(permissions.allObjects, nil);
+    }];
 }
 
 - (void)applyPermission:(RLMSyncPermission *)permission callback:(RLMPermissionStatusBlock)callback {
-    verifyInRunLoop();
-    if (!_user || _user->state() == SyncUser::State::Error) {
-        callback(make_permission_error_change(@"Permissions cannot be applied using an invalid user."));
+    if (NSError *error = checkUser(_user, @"applied")) {
+        callback(error);
         return;
     }
-    Permissions::set_permission(_user,
-                                [permission rawPermission],
-                                RLMWrapPermissionStatusCallback(callback),
-                                *_configMaker);
+    id condition;
+    if (permission.identity) {
+        condition = @{@"userId": permission.identity};
+    }
+    else {
+        condition = @{@"metadataKey": permission.key, @"metadataValue": permission.value};
+    }
+    [RLMSyncApplyPermissionsEndpoint
+     sendRequestToServer:self.authenticationServer
+     JSON:@{kRLMSyncTokenKey: self.refreshToken,
+            @"condition": condition,
+            @"realmPath": permission.path,
+            @"accessLevel": RLMSyncAccessLevelToString(permission.accessLevel)}
+     completion:callback];
 }
 
-- (void)revokePermission:(RLMSyncPermission *)permission callback:(RLMPermissionStatusBlock)callback {
-    verifyInRunLoop();
-    if (!_user || _user->state() == SyncUser::State::Error) {
-        callback(make_permission_error_change(@"Permissions cannot be revoked using an invalid user."));
-        return;
-    }
-    Permissions::delete_permission(_user,
-                                   [permission rawPermission],
-                                   RLMWrapPermissionStatusCallback(callback),
-                                   *_configMaker);
+static id ISO8601Formatter() {
```

Comment:
[q] Why not return `NSDateFormatter`? Also, could this be in a utility class instead of on the SyncUser?

Reply:
It was conditionally returning NSISO8601DateFormatter if running on an OS version which supports it, but that turned out not to work and I forgot to change the return type. I guess I might as well toss it in RLMUtil.